### PR TITLE
[Snappi] Adding Lossless Response to Throttling Pause Storms Testcase

### PR DESCRIPTION
**Pull request by dks0692**
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/sonic-net/SONiC/blob/gh-pages/CONTRIBUTING.md

Please provide following information to help code review process a bit easier:
-->
### Description of PR
Tests the reduction in lossless streams bandwidth when subject to throttling PFC pause messages.
<!--
This test tests the reduction in lossless streams bandwidth when subject to throttling PFC pause messages.
-->

Summary:
Fixes # (issue)

### Type of change

<!--
- Fill x for your type of change.
- e.g.
- [x] Bug fix
-->

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)


### Back port request
- [ ] 201911
- [ ] 202012
- [ ] 202205

### Approach
#### What is the motivation for this PR?
To test the reduction in lossless streams bandwidth when subject to throttling PFC pause messages. The PFC watchdog, which should be enabled for this test, should not intervene.

#### How did you do it?
It was tested by having back ground and test flow @ 100% line rate on the rx port and rx port sending the PFC pause frames to the tx side thereby causing partial loss on the test flow traffic
#### How did you verify/test it?
Tested against voq chassis with SONiC installed in it
#### Any platform specific information?

#### Supported testbed topology if it's a new test case?
T2, T1 & T0
### Documentation
Testplan - https://github.com/sonic-net/sonic-mgmt/pull/8373

<!--
(If it's a new feature, new test case)
Did you update documentation/Wiki relevant to your implementation?
Link to the wiki page?
-->
### Output

AzDevOps@6105354dad99:~/sonic-mgmt/tests$ py.test --inventory ../ansible/snappi-sonic --host-pattern sonic-s6100-dut --testbed vms-snappi-sonic --testbed_file ../ansible/testbed.csv --show-capture=stdout --log-cli-level info --showlocals -ra --allow_recover --skip_sanity --disable_loganalyzer snappi_tests/multidut/pfc/test_lossless_response_to_throttling_pause_storms.py 
/usr/local/lib/python2.7/dist-packages/ansible/parsing/vault/__init__.py:44: CryptographyDeprecationWarning: Python 2 is no longer supported by the Python core team. Support for it is now deprecated in cryptography, and will be removed in the next release.
  from cryptography.exceptions import InvalidSignature
============================================================================================================== test session starts ==============================================================================================================
platform linux2 -- Python 2.7.18, pytest-4.6.11, py-1.11.0, pluggy-0.13.1
ansible: 2.8.20
rootdir: /var/AzDevOps/sonic-mgmt/tests, inifile: pytest.ini
plugins: allure-pytest-2.8.22, celery-4.4.7, ansible-2.2.4, forked-1.3.0, html-1.22.1, metadata-1.11.0, repeat-0.9.1, xdist-1.28.0
collecting 1 item                                                                                                                                                                                                                               
-------------------------------------------------------------------------------------------------------------- live log collection --------------------------------------------------------------------------------------------------------------
22:09:58 __init__.pytest_collection_modifyitems   L0582 INFO   | Available basic facts that can be used in conditional skip:
{
  "topo_type": "ptf", 
  "testbed": "vms-snappi-sonic", 
  "topo_name": "ptf64"
}

---------------------------------------------------------------------------------------------------------------- live log setup -----------------------------------------------------------------------------------------------------------------
22:09:58 __init__.set_default                     L0054 INFO   | Completeness level not set during test execution. Setting to default level: CompletenessLevel.basic
22:09:58 __init__.check_test_completeness         L0152 INFO   | Test has no defined levels. Continue without test completeness checks
22:10:04 ptfhost_utils.run_icmp_responder_session L0239 INFO   | Skip running icmp_responder at session level, it is only for dualtor testbed with active-active mux ports.
22:10:04 conftest.creds_on_dut                    L0704 INFO   | dut sonic-s6100-dut belongs to groups [u'snappi-sonic', u'sonic', u'sonic_dell64_40', 'fanout']
22:10:04 conftest.creds_on_dut                    L0728 INFO   | skip empty var file ../ansible/group_vars/all/corefile_uploader.yml
22:10:04 conftest.creds_on_dut                    L0728 INFO   | skip empty var file ../ansible/group_vars/all/env.yml
22:10:04 conftest.nbrhosts                        L0521 INFO   | No VMs exist for this topology: ptf64
22:10:04 conftest.core_dump_and_config_check      L2032 INFO   | Collecting core dumps before test on sonic-s6100-dut
22:10:05 conftest.core_dump_and_config_check      L2041 INFO   | Collecting running config before test on sonic-s6100-dut
22:10:10 __init__.sanity_check                    L0125 INFO   | Skip sanity check according to command line argument
22:10:10 conftest.generate_params_dut_hostname    L1108 INFO   | Using DUTs ['sonic-s6100-dut'] in testbed 'vms-snappi-sonic'
22:10:10 conftest.rand_one_dut_hostname           L0378 INFO   | Randomly select dut sonic-s6100-dut for testing
22:10:10 __init__._fixture_generator_decorator    L0081 INFO   | -------------------- fixture enable_packet_aging_after_test setup starts --------------------
22:10:10 __init__._fixture_generator_decorator    L0085 INFO   | -------------------- fixture enable_packet_aging_after_test setup ends --------------------
22:10:10 __init__._fixture_generator_decorator    L0081 INFO   | -------------------- fixture rand_lossless_prio setup starts --------------------
22:10:10 __init__._fixture_generator_decorator    L0085 INFO   | -------------------- fixture rand_lossless_prio setup ends --------------------
22:10:10 __init__._fixture_generator_decorator    L0081 INFO   | -------------------- fixture rand_lossy_prio setup starts --------------------
22:10:10 __init__._fixture_generator_decorator    L0085 INFO   | -------------------- fixture rand_lossy_prio setup ends --------------------
22:10:10 __init__._fixture_generator_decorator    L0081 INFO   | -------------------- fixture start_pfcwd_after_test setup starts --------------------
22:10:10 __init__._fixture_generator_decorator    L0085 INFO   | -------------------- fixture start_pfcwd_after_test setup ends --------------------
22:10:10 __init__._fixture_func_decorator         L0069 INFO   | -------------------- fixture snappi_api_serv_ip setup starts --------------------
22:10:10 __init__._fixture_func_decorator         L0076 INFO   | -------------------- fixture snappi_api_serv_ip setup ends --------------------
22:10:10 __init__._fixture_func_decorator         L0069 INFO   | -------------------- fixture snappi_api_serv_port setup starts --------------------
22:10:10 __init__._fixture_func_decorator         L0076 INFO   | -------------------- fixture snappi_api_serv_port setup ends --------------------
22:10:10 __init__._fixture_generator_decorator    L0081 INFO   | -------------------- fixture snappi_api setup starts --------------------
22:10:10 __init__._fixture_generator_decorator    L0085 INFO   | -------------------- fixture snappi_api setup ends --------------------
22:10:10 __init__._fixture_func_decorator         L0069 INFO   | -------------------- fixture prio_dscp_map setup starts --------------------
22:10:12 __init__._fixture_func_decorator         L0076 INFO   | -------------------- fixture prio_dscp_map setup ends --------------------
22:10:12 __init__._fixture_func_decorator         L0069 INFO   | -------------------- fixture lossless_prio_list setup starts --------------------
22:10:13 __init__._fixture_func_decorator         L0076 INFO   | -------------------- fixture lossless_prio_list setup ends --------------------
22:10:13 __init__.loganalyzer                     L0045 INFO   | Log analyzer is disabled
22:10:13 __init__._fixture_func_decorator         L0069 INFO   | -------------------- fixture get_multidut_snappi_ports setup starts --------------------
22:10:13 __init__._fixture_func_decorator         L0076 INFO   | -------------------- fixture get_multidut_snappi_ports setup ends --------------------
----------------------------------------------------------------------------------------------------------------- live log call -----------------------------------------------------------------------------------------------------------------
22:10:13 snappi_fixtures.__intf_config_multidut   L0732 INFO   | Configuring Dut: sonic-s6100-dut with port Ethernet12 with IP 20.0.1.1/24
22:10:14 snappi_fixtures.__intf_config_multidut   L0732 INFO   | Configuring Dut: sonic-s6100-dut with port Ethernet8 with IP 20.0.2.1/24
22:10:16 snappi_fixtures.__intf_config_multidut   L0732 INFO   | Configuring Dut: sonic-s6100-dut with port Ethernet0 with IP 20.0.3.1/24
22:10:22 connection._warn                         L0246 WARNING| Verification of certificates is disabled
22:10:22 connection._info                         L0243 INFO   | Determining the platform and rest_port using the 10.36.78.135 address...
22:10:22 connection._warn                         L0246 WARNING| Unable to connect to http://10.36.78.135:443.
22:10:22 connection._info                         L0243 INFO   | Connection established to `https://10.36.78.135:443 on linux`
22:10:36 connection._info                         L0243 INFO   | Using IxNetwork api server version 9.30.2212.7
22:10:36 connection._info                         L0243 INFO   | User info IxNetwork/ixnetworkweb/admin-52-14494
22:10:37 snappi_api.info                          L1132 INFO   | snappi-0.9.1
22:10:37 snappi_api.info                          L1132 INFO   | snappi_ixnetwork-0.9.1
22:10:37 snappi_api.info                          L1132 INFO   | ixnetwork_restpy-1.0.64
22:10:37 snappi_api.info                          L1132 INFO   | Config validation 0.011s
22:10:39 snappi_api.info                          L1132 INFO   | Ports configuration 1.733s
22:10:39 snappi_api.info                          L1132 INFO   | Captures configuration 0.160s
22:10:42 snappi_api.info                          L1132 INFO   | Add location hosts [10.36.78.53] 2.372s
22:10:46 snappi_api.info                          L1132 INFO   | Location hosts ready [10.36.78.53] 4.174s
22:10:47 snappi_api.info                          L1132 INFO   | Speed conversion is not require for (port.name, speed) : [('Port 0', u'novusHundredGigNonFanOut'), ('Port 1', u'novusHundredGigNonFanOut'), ('Port 2', u'novusHundredGigNonFanOut')]
22:10:47 snappi_api.info                          L1132 INFO   | Aggregation mode speed change 0.436s
22:10:53 snappi_api.info                          L1132 INFO   | Location preemption [10.36.78.53;6;4, 10.36.78.53;6;3, 10.36.78.53;6;1] 0.119s
22:11:19 snappi_api.info                          L1132 INFO   | Location connect [Port 0, Port 1, Port 2] 25.056s
22:11:19 snappi_api.warning                       L1138 WARNING| Port 0 connectedLinkDown
22:11:19 snappi_api.warning                       L1138 WARNING| Port 1 connectedLinkDown
22:11:19 snappi_api.warning                       L1138 WARNING| Port 2 connectedLinkDown
22:11:19 snappi_api.info                          L1132 INFO   | Location state check [Port 0, Port 1, Port 2] 0.387s
22:11:19 snappi_api.info                          L1132 INFO   | Location configuration 39.519s
22:11:37 snappi_api.info                          L1132 INFO   | Layer1 configuration 17.642s
22:11:37 snappi_api.info                          L1132 INFO   | Lag Configuration 0.083s
22:11:37 snappi_api.info                          L1132 INFO   | Convert device config : 0.233s
22:11:37 snappi_api.info                          L1132 INFO   | Create IxNetwork device config : 0.000s
22:11:38 snappi_api.info                          L1132 INFO   | Push IxNetwork device config : 0.679s
22:11:38 snappi_api.info                          L1132 INFO   | Devices configuration 0.985s
22:11:42 snappi_api.info                          L1132 INFO   | Flows configuration 4.255s
22:11:50 snappi_api.info                          L1132 INFO   | Start interfaces 8.251s
22:11:51 snappi_api.info                          L1132 INFO   | IxNet - The Traffic Item was modified. Please perform a Traffic Generate to update the associated traffic Flow Groups
22:11:51 traffic_generation.run_traffic           L0548 INFO   | Wait for Arp to Resolve ...
22:11:52 traffic_generation.run_traffic           L0566 INFO   | Starting transmit on all flows ...
22:11:56 snappi_api.info                          L1132 INFO   | Flows generate/apply 3.081s
22:12:08 snappi_api.info                          L1132 INFO   | Flows clear statistics 11.979s
22:12:08 snappi_api.info                          L1132 INFO   | Captures start 0.000s
22:12:12 snappi_api.info                          L1132 INFO   | Flows start 3.617s
22:12:12 traffic_generation.run_traffic           L0573 INFO   | Polling DUT for traffic statistics for 20 seconds ...
22:12:32 traffic_generation.run_traffic           L0591 INFO   | DUT polling complete
22:12:47 traffic_generation.run_traffic           L0607 INFO   | All test and background traffic flows stopped
22:12:49 traffic_generation.run_traffic           L0630 INFO   | Dumping per-flow statistics
22:12:50 traffic_generation.run_traffic           L0634 INFO   | Stopping transmit on all remaining flows
22:12:51 traffic_generation.verify_m2o_results    L1053 INFO   | Test Flow 1 -> 0, TX Frames:71839080, RX Frames:71835326
22:12:51 traffic_generation.verify_m2o_results    L1053 INFO   | Test Flow 2 -> 0, TX Frames:71839080, RX Frames:71835322
22:12:51 traffic_generation.verify_m2o_results    L1046 INFO   | Background Flow 1 -> 0, TX Frames:59865900, RX Frames:59865900
22:12:51 traffic_generation.verify_m2o_results    L1046 INFO   | Background Flow 2 -> 0, TX Frames:59865900, RX Frames:59865900
22:12:51 snappi_fixtures.cleanup_config           L0885 INFO   | Removing Configuration on Dut: sonic-s6100-dut with port Ethernet12 with ip :20.0.1.1/24
22:12:53 snappi_fixtures.cleanup_config           L0885 INFO   | Removing Configuration on Dut: sonic-s6100-dut with port Ethernet8 with ip :20.0.2.1/24
22:12:54 snappi_fixtures.cleanup_config           L0885 INFO   | Removing Configuration on Dut: sonic-s6100-dut with port Ethernet0 with ip :20.0.3.1/24
PASSED                                                                                                                                                                                                                                    [100%]
--------------------------------------------------------------------------------------------------------------- live log teardown ---------------------------------------------------------------------------------------------------------------
22:12:56 __init__._fixture_generator_decorator    L0093 INFO   | -------------------- fixture snappi_api teardown starts --------------------
22:13:13 __init__._fixture_generator_decorator    L0102 INFO   | -------------------- fixture snappi_api teardown ends --------------------
22:13:13 __init__._fixture_generator_decorator    L0093 INFO   | -------------------- fixture start_pfcwd_after_test teardown starts --------------------
22:13:14 __init__._fixture_generator_decorator    L0102 INFO   | -------------------- fixture start_pfcwd_after_test teardown ends --------------------
22:13:14 __init__._fixture_generator_decorator    L0093 INFO   | -------------------- fixture rand_lossy_prio teardown starts --------------------
22:13:14 __init__._fixture_generator_decorator    L0102 INFO   | -------------------- fixture rand_lossy_prio teardown ends --------------------
22:13:14 __init__._fixture_generator_decorator    L0093 INFO   | -------------------- fixture rand_lossless_prio teardown starts --------------------
22:13:14 __init__._fixture_generator_decorator    L0102 INFO   | -------------------- fixture rand_lossless_prio teardown ends --------------------
22:13:14 __init__._fixture_generator_decorator    L0093 INFO   | -------------------- fixture enable_packet_aging_after_test teardown starts --------------------
22:13:16 __init__._fixture_generator_decorator    L0102 INFO   | -------------------- fixture enable_packet_aging_after_test teardown ends --------------------
22:13:16 conftest.core_dump_and_config_check      L2069 INFO   | Collecting core dumps after test on sonic-s6100-dut
22:13:17 conftest.core_dump_and_config_check      L2086 INFO   | Collecting running config after test on sonic-s6100-dut
22:13:18 conftest.core_dump_and_config_check      L2211 WARNING| Core dump or config check failed for test_lossless_response_to_throttling_pause_storms.py, results: {"core_dump_check": {"new_core_dumps": {"sonic-s6100-dut": []}, "pass": true}, "config_db_check": {"cur_only_config": {"sonic-s6100-dut": {"null": {"INTERFACE": {"Ethernet8": {}, "Ethernet4|20.0.2.1/24": {}, "Ethernet4": {}, "Ethernet8|20.0.1.1/24": {}, "Ethernet4|20.0.1.1/24": {}, "Ethernet12|20.0.2.1/24": {}, "Ethernet8|20.0.3.1/24": {}, "Ethernet12": {}}}}}, "inconsistent_config": {"sonic-s6100-dut": {"null": {}}}, "pre_only_config": {"sonic-s6100-dut": {"null": {"VLAN_MEMBER": {"Vlan2|Ethernet8": {"tagging_mode": "untagged"}, "Vlan2|Ethernet0": {"tagging_mode": "untagged"}, "Vlan2|Ethernet12": {"tagging_mode": "untagged"}, "Vlan2|Ethernet4": {"tagging_mode": "untagged"}}}}}, "pass": false}}


=============================================================================================================== warnings summary ================================================================================================================
/usr/local/lib/python2.7/dist-packages/_pytest/config/__init__.py:545
  /usr/local/lib/python2.7/dist-packages/_pytest/config/__init__.py:545: PytestAssertRewriteWarning: Module already imported so cannot be rewritten: tests.common.plugins.loganalyzer
    self.import_plugin(import_spec)

/usr/local/lib/python2.7/dist-packages/_pytest/config/__init__.py:545
  /usr/local/lib/python2.7/dist-packages/_pytest/config/__init__.py:545: PytestAssertRewriteWarning: Module already imported so cannot be rewritten: tests.common.plugins.sanity_check
    self.import_plugin(import_spec)

/usr/local/lib/python2.7/dist-packages/_pytest/config/__init__.py:545
  /usr/local/lib/python2.7/dist-packages/_pytest/config/__init__.py:545: PytestAssertRewriteWarning: Module already imported so cannot be rewritten: tests.common.dualtor
    self.import_plugin(import_spec)

snappi_tests/dinesh_response_time_pr/test_lossless_response_to_throttling_pause_storms.py::test_lossless_response_to_throttling_pause_storms[linecard_configuration_set0-non_chassis_single_line_card]
  /usr/local/lib/python2.7/dist-packages/pytest_ansible/module_dispatcher/v28.py:81: UserWarning: provided hosts list is empty, only localhost is available
    warnings.warn("provided hosts list is empty, only localhost is available")

-- Docs: https://docs.pytest.org/en/latest/warnings.html
------------------------------------------------------------------------------------------------------------ live log sessionfinish -------------------------------------------------------------------------------------------------------------
22:13:18 __init__.pytest_terminal_summary         L0064 INFO   | Can not get Allure report URL. Please check logs
==================== 1 passed, 4 warnings in 201.21 seconds ====================
INFO:root:Can not get Allure report URL. Please check logs

### CODE REVIEW

**Comment by dks0692** [comment]
@developfast Can you please review

**Review by developfast — CHANGES_REQUESTED** [review]

File: tests/snappi_tests/multidut/pfc/files/lossless_response_to_throttling_pause_storms_helper.py [hunk 1]
```diff
@@ -0,0 +1,391 @@
...
+BG_FLOW_AGGR_RATE_PERCENT = 25
+PAUSE_FLOW_RATE = 15
+DATA_PKT_SIZE = 1024
+DATA_FLOW_DURATION_SEC = 20
+DATA_FLOW_DELAY_SEC = 0
+SNAPPI_POLL_DELAY_SEC = 2
+TOLERANCE_THRESHOLD = 0.05
+PAUSE_FLOW_DURATION_SEC = 10
+PAUSE_FLOW_DELAY_SEC = 5
+
```
Comment (vmittal-msft) [inline tests/snappi_tests/multidut/pfc/files/lossless_response_to_throttling_pause_storms_helper.py:35]: please update how the test is done and expected results.

File: tests/snappi_tests/multidut/pfc/test_lossless_response_to_throttling_pause_storms.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+                                                      fanout_graph_facts,                # noqa: F811
+                                                      line_card_choice,
+                                                      duthosts,
+                                                      prio_dscp_map,                    # noqa: F811
+                                                      lossless_prio_list,               # noqa: F811
+                                                      linecard_configuration_set,       # noqa: F811
+                                                      get_multidut_snappi_ports,):      # noqa: F811
+
+    """
+    Run PFC watchdog test under many to one traffic pattern
```
Comment (vmittal-msft) [inline tests/snappi_tests/multidut/pfc/test_lossless_response_to_throttling_pause_storms.py:32]: please fix comments.

File: tests/snappi_tests/multidut/pfc/test_lossless_response_to_throttling_pause_storms.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        snappi_api (pytest fixture): SNAPPI session
+        snappi_testbed_config (pytest fixture): testbed configuration information
+        conn_graph_facts (pytest fixture): connection graph
+        fanout_graph_facts (pytest fixture): fanout graph
+        duthosts (pytest fixture): list of DUTs
+        lossy_prio_list (pytest fixture): list of lossy priorities
+        prio_dscp_map (pytest fixture): priority vs. DSCP map (key = priority)
+        line_card_choice: Line card choice to be mentioned in the variable.py file
+        linecard_configuration_set : Line card classification, (min 1 or max 2  hostnames and asics to be given)
+
```
Comment (vmittal-msft) [inline tests/snappi_tests/multidut/pfc/test_lossless_response_to_throttling_pause_storms.py:44]: Do we expect loss to happen or just traffic reduction ? I believe it should just reduce traffic and no loss for lossless traffic.

File: tests/snappi_tests/multidut/pfc/test_lossless_response_to_throttling_pause_storms.py (hunk 2 above)
Comment (dks0692) [reply tests/snappi_tests/multidut/pfc/test_lossless_response_to_throttling_pause_storms.py:32]: fixed

File: tests/snappi_tests/multidut/pfc/files/lossless_response_to_throttling_pause_storms_helper.py (hunk 1 above)
Comment (dks0692) [reply tests/snappi_tests/multidut/pfc/files/lossless_response_to_throttling_pause_storms_helper.py:35]: The background traffic will include two lossy traffic streams, each with randomly chosen priorities (0..2, 5..7), and each having a 25% bandwidth. The test data traffic will consist of two lossless traffic streams, with the SONiC default lossless priorities of 3 and 4, and each having a 25% bandwidth. PFC pause throttling stream: Persistent PFC pause frames from the IXIA Rx port with enough repetition and quanta chosen so as to reduce one or all lossless streams down to 90% of their configured bandwidth.Start the background and test traffic from IXIA Tx ports 1 & 2; the combined line rate will be 100% of the egress port capacity (25% line rate bandwidth per stream)
After a few seconds, from IXIA port 3, start the PFC throttling stream onto one or all lossless priorities. Ensure that: a. The IXIA RX port indicates that the targeted lossless streams got their bandwidth reduced by 90% b. The IXIA Rx port is receiving all lossy streams and any of the lossless streams not marked for PFC pause to their nominal configured transmit bandwidth

**Comment by selldinesh** [comment]
@yxieca @wangxin please help merge.

**Review by vmittal-msft — APPROVED** [review]

**Comment by vmittal-msft** [comment]
@developfast  please see if good from your side.

**Comment by vmittal-msft** [comment]
@yxieca @developfast Can you please help merge ?

**Comment by vmittal-msft** [comment]
@yxieca @wangxin please help merge as tests passing now

**Comment by yxieca** [comment]
@developfast  are we ready to merge this PR?

**Pull request merged**